Cesium3DTilesSelection/src/TilesetContentManager.cpp

```diff
@@ -0,0 +1,361 @@
+
+        notifyTileDoneLoading(tile);
+      })
+      .catchInMainThread(
+          [pLogger = _externals.pLogger, &tile, this](std::exception&& e) {
+            notifyTileDoneLoading(tile);
+            SPDLOG_LOGGER_ERROR(
+                pLogger,
+                "An unexpected error occurs when loading tile: {}",
+                e.what());
+          });
+}
+
+void TilesetContentManager::updateTileContent(Tile& tile) {
+  const TileContent& content = tile.getContent();
+
+  TileLoadState state = content.getState();
+  switch (state) {
+  case TileLoadState::ContentLoaded:
+    updateContentLoadedState(tile);
+    break;
+  default:
+    break;
+  }
+}
+
+bool TilesetContentManager::unloadTileContent(Tile& tile) {
+  TileContent& content = tile.getContent();
+  TileLoadState state = content.getState();
+  if (state == TileLoadState::Unloaded) {
+    return true;
+  }
+
+  if (state == TileLoadState::ContentLoading) {
+    return false;
+  }
+
+  if (content.isExternalContent() || content.isEmptyContent()) {
+    return false;
+  }
+
+  notifyTileUnloading(tile);
+
+  switch (state) {
+  case TileLoadState::ContentLoaded:
+    unloadContentLoadedState(tile);
+    break;
+  case TileLoadState::Done:
+    unloadDoneState(tile);
+    break;
+  default:
+    break;
+  }
+
+  content.setContentKind(TileUnknownContent{});
+  content.setTileInitializerCallback({});
+  content.setState(TileLoadState::Unloaded);
+  return true;
+}
+
+const std::vector<CesiumAsync::IAssetAccessor::THeader>&
+TilesetContentManager::getRequestHeaders() const noexcept {
+  return _requestHeaders;
+}
+
+std::vector<CesiumAsync::IAssetAccessor::THeader>&
+TilesetContentManager::getRequestHeaders() noexcept {
+  return _requestHeaders;
+}
+
+int32_t TilesetContentManager::getNumOfTilesLoading() const noexcept {
+  return _tilesLoadOnProgress;
+}
+
+int64_t TilesetContentManager::getTilesDataUsed() const noexcept {
+  return _tilesDataUsed;
+}
+
+void TilesetContentManager::setTileContent(
+    TileContent& content,
+    TileLoadResult&& result,
+    void* pWorkerRenderResources) {
+  switch (result.state) {
+  case TileLoadResultState::Success:
+    content.setState(TileLoadState::ContentLoaded);
+    break;
+  case TileLoadResultState::Failed:
+    content.setState(TileLoadState::Failed);
+    break;
+  case TileLoadResultState::RetryLater:
+    content.setState(TileLoadState::FailedTemporarily);
+    break;
+  default:
+    assert(false && "Cannot handle an unknown TileLoadResultState");
+    break;
+  }
+
+  content.setContentKind(std::move(result.contentKind));
+  content.setTileInitializerCallback(std::move(result.tileInitializer));
+  content.setRenderResources(pWorkerRenderResources);
+}
+
+void TilesetContentManager::updateContentLoadedState(Tile& tile) {
+  // initialize this tile content first
+  TileContent& content = tile.getContent();
+  auto& tileInitializer = content.getTileInitializerCallback();
+  if (tileInitializer) {
+    tileInitializer(tile);
```

Comment:
Can we clear the tileInitializer after invoking it, so its resources can be freed?